File: src/main/java/github/macro/build_info/gems/Acquisition.java
package github.macro.build_info.gems;

import java.util.List;

/**
 * Created by Macro303 on 2020-Jan-09.
 */
public class Acquisition {
	private final List<Recipe> recipes;
	private final List<Quest> quests;
	private final List<Vendor> vendors;

	public Acquisition(List<Recipe> recipes, List<Quest> quests, List<Vendor> vendors) {
		this.recipes = recipes;
		this.quests = quests;
		this.vendors = vendors;
	}

	public List<Recipe> getRecipes() {
		return recipes;
	}

	public List<Quest> getQuests() {
		return quests;
	}

	public List<Vendor> getVendors() {
		return vendors;
	}

	@Override
	public String toString() {
		return "Acquisition{" +
				"recipes=" + recipes +
				", quests=" + quests +
				", vendors=" + vendors +
				'}';
	}
}

File: src/main/kotlin/github/macro/build_info/gems/Acquisition.kt
package github.macro.build_info.gems

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.deser.std.StdDeserializer
import github.macro.Util
import github.macro.build_info.gems.ingredient.Ingredient
import github.macro.build_info.gems.reward.Reward
import javafx.beans.property.SimpleListProperty
import javafx.collections.FXCollections
import tornadofx.*
import java.io.IOException

/**
 * Created by Macro303 on 2020-Jan-13.
 */
@JsonDeserialize(using = AcquisitionDeserializer::class)
class Acquisition(
	rewards: List<Reward>,
	crafting: List<List<Ingredient>>
) {
	val rewardsProperty = SimpleListProperty<Reward>()
	var rewards by rewardsProperty

	val craftingProperty = SimpleListProperty<List<Ingredient>>()
	var crafting by craftingProperty

	init {
		this.rewards = FXCollections.observableList(rewards)
		this.crafting = FXCollections.observableList(crafting)
	}

	override fun toString(): String {
		return "Acquisition(rewardsProperty=$rewardsProperty, craftingProperty=$craftingProperty)"
	}
}

class AcquisitionDeserializer @JvmOverloads constructor(vc: Class<*>? = null) : StdDeserializer<Acquisition?>(vc) {

	@Throws(IOException::class, JsonProcessingException::class)
	override fun deserialize(parser: JsonParser, ctx: DeserializationContext): Acquisition? {
		val node: JsonNode = parser.readValueAsTree()

		val rewards = node["rewards"].map { Util.JSON_MAPPER.treeToValue(it, Reward::class.java) }
		val crafting = node["crafting"].map { ingredients ->
			ingredients.map { Util.JSON_MAPPER.treeToValue(it, Ingredient::class.java) }
		}

		return Acquisition(rewards = rewards, crafting = crafting)
	}
}